File: 0064-minimum-path-sum/0064-minimum-path-sum.rs
impl Solution {
    pub fn min_path_sum(grid: Vec<Vec<i32>>) -> i32 {
        if grid.is_empty() || grid[0].is_empty() {
            return 0;
        }
        
        let n = grid.len();
        let m = grid[0].len();
        
        let mut dp = vec![vec![0; m]; n];
        
        dp[0][0] = grid[0][0];
        
        for i in 1..m {
            dp[0][i] = dp[0][i - 1] + grid[0][i];
        }
        
        for j in 1..n {
            dp[j][0] = dp[j - 1][0] + grid[j][0];
        }
        
        for i in 1..n {
            for j in 1..m {
                dp[i][j] = std::cmp::min(dp[i - 1][j], dp[i][j - 1]) + grid[i][j];
            }
        }
        
        dp[n - 1][m - 1]
    }
}

File: 0064-minimum-path-sum/0064-minimum-path-sum.kt
class Solution {
    fun minPathSum(grid: Array<IntArray>): Int {
        val m = grid.size
        val n = grid[0].size
        val memo = Array(m) { IntArray(n) { -1 } }

        fun helper(x: Int, y: Int): Int {
            if (x == m - 1 && y == n - 1) {
                return grid[x][y]
            }
            if (x == m - 1) {
                return grid[x][y] + helper(x, y + 1)
            }
            if (y == n - 1) {
                return grid[x][y] + helper(x + 1, y)
            }

            if (memo[x][y] != -1) {
                return memo[x][y]
            }

            val down = helper(x + 1, y)
            val right = helper(x, y + 1)

            memo[x][y] = minOf(down, right) + grid[x][y]
            return memo[x][y]
        }

        return helper(0, 0)
    }
}
